src/stdhasher.cr

```diff
@@ -0,0 +1,248 @@
+require "secure_random"
+
+# Hasher usable for `def hash(hasher)` should satisfy protocol:
+#   class MyHasher
+#     def <<(v : Int::Primitive)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(b : Bytes)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(n : Nil)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(v)
+#       v.hash(self)
+#       self
+#     end
+#
+#     # digest returns hashsum for current state without state mutation
+#     # Note: hashsum should implement commutative `+` if MyHasher used
+#     # for `Hasher#hash(hasher)`
+#     def digest
+#     end
+#
+#     def clone_build
+#       with_state_copy do |copy|
+#         yield copy
+#         copy.digest
+#       end
+#     end
+#   end
+
+# StdHasher used as standard hasher in `Object#hash`
+# It have to provide defenense against HashDos, and be reasonably fast.
+# To protect against HashDos, it is seeded with secure random, and have
+# permutation that hard to forge without knowing seed and seeing hash digest.
+#
+# Also it has specialized methods for primitive keys with different seeds.
+struct StdHasher
+  @@seed = StaticArray(UInt32, 7).new{|i| 0_u32}
+  def self.init
+    raise "#{@@seed[0]} #{@@seed[1]}" unless @@seed[0] == 0_u32 && @@seed[1] == 0_u32
+    buf = pointerof(@@seed).as(Pointer(UInt8))
+    SecureRandom.random_bytes(buf.to_slice(sizeof(typeof(@@seed))))
+  end
+
+  private def initialize(@h : Pointer(Impl))
+  end
+
+  def initialize
+    @h = Pointer(Impl).malloc(1, Impl.new(@@seed[0], @@seed[1]))
```

Comment:
I never said I want HEAP allocations (thus switch to classes): it's noticeably slower (bad) and pushes pressure to GC (bad).

What I don't like is passing stack-pointers around, especially to uncontrolled contexts. This is an anti-pattern in Crystal and I'd like to avoid it. For example we prefer `Slice` over `Pointer`, `to_slice` over `to_unsafe`, unless we're in a controlled context (e.g. use `unsafe_at(i)` when `i` is already constrained). I wish #124 was already implemented so anything that deals with pointers was harder to use, not trivial.

If we can't find a better solution, then it's acceptable, but I'd still like a better solution. For example I like the idea of yielding values to a digest: no need to pass a struct around, no pointers, ...

BTW: you're pushing your example too far, Hasher can mutate itself (as demonstrated above). The only requirement is it must be returned, since it was passed by copy:

```crystal
module Indexable
  def hash(hasher)
    hasher.raw(size
    each { |elem| hasher << elem }
    hasher
  end
end
```